### DISCUSSION TITLE
Point-Stat ORANK output

### DISCUSSION THREAD

**Discussion by jwraby** [dtcenter/METplus#1539] [Statistical Computation] [answered]
I ran MET V10.1.0 Point-Stat with the HiRA option for a range of neighborhood sizes for the TMP surface variable (Z2). I'm trying to understand the contents of the ORANK output text file. For each neighborhood size (NBRHD_SQUARE) there is a corresponding total number of observations (TOTAL) and number of ensemble members (N_ENS). Each row of data in the file for a given neighborhood size is for a particular observation such that the number of rows is equal to the TOTAL value. Looking at the relationship among neighborhood size, number of observations, and number of ensemble members, I assembled some data in a text file which shows the summary data for the above values taken from the ORANK file. One pattern which emerges is that the smallest neighborhood has the maximum TOTAL value, but the minimum number of ensemble members. The largest neighborhood has the minimum TOTAL value, but the maximum number of ensemble members. What is causing the reduction in the TOTAL value (by about 50% relative to that for the smallest neighborhood size) for the largest neighborhood size and why does the number of members increase by a few orders of magnitude over that for the smallest neighborhood size? Thanks. 

PS: I cannot get the 2 kb text (.txt) file to upload using drag/drop or browse to file functionality. Reports "Something went really wrong, and we can't process that file". I will try uploading after clicking on "Start discussion".

**Comment by JohnHalleyGotway** [comment, marked as answer]
@jwraby I see you have some questions about the HiRA output computed by Point-Stat... and the ORANK line type in particular. You've noticed 2 patterns. As the neighborhood size increases...
1. The number of ensemble members in the N_ENS columns also increases.
2. The number of matched pairs in the TOTAL column decreases.

The behavior of (1.) is pretty obvious and expected. As the neighborhood size increases, more nearby grid points are used and the size of the "spatial ensemble" increases. When the shape is SQUARE, the N_ENS is just the width x width. When the shape is CIRCLE, N_ENS is something less than width x width.

The behavior of (2.) is a little less obvious. Put simply, it's the impact of edge effects and the valid data threshold (vld_thresh) setting. I realize that we're talking about Point-Stat but looking at the NetCDF matched pairs output from Grid-Stat demonstrates an important point.

The image below shows the output from 4 runs of plot data plane. The top-left is 2-m Temperature. The top-right is that field, thresholded > 300 and smoothed with a 3 x 3 neighborhood size. Bottom-left uses a 5 x 5 neighborhood and bottom-right 11 x 11:
<img width="635" alt="Screen Shot 2022-03-29 at 9 50 41 AM" src="https://user-images.githubusercontent.com/21087144/160653025-7aa09515-097c-489b-8cdb-e86f5e76bcad.png">

Note that as we increase the width from 3 to 5 to 11, the extent of valid data shrinks. For example, you can see how the amount of Cuba that's included shrinks.

As the neighborhood size increases, the number of points included increases. But along the edge of the region of valid data, we're including more grid point that contain bad data values. This run uses "nbrhd.vld_thresh = 1.0". That requires that ALL the neighborhood points contain valid values in order to compute a fractional coverage value. With this setting, anytime the neighborhood extends into bad data or off the grid, you'd get bad data in the output.

So what should you do about it?

While I illustrated with Grid-Stat output, HiRA in Point-Stat isn't very flexible. If ANY of the "ensemble members" (i.e. neighborhood points) contains bad data, then that point is discarded. The reason for that is that there's no obvious way how to compute an observation rank value when bad data is present. This effectively means that Point-Stat uses vld_thresh = 1.0.

You could use a masking region to restrict your analysis to the interior of the domain so that you don't use any observations that are with N points of the edge, where N is the maximum neighborhood size you're using.

That should result in the TOTAL column remaining constant across all neighborhood sizes.

Hope that helps clarify!

**Reply by jwraby** [reply]
Thanks for your very helpful explanation. Regarding 2, for me, your explanation has clarified something I wasn't understanding about how neighborhoods are applied. If I am understanding this correctly, increasing the size of the neighborhoods causes them to include areas outside the boundaries of the model domain and resulting in the inclusion of increasing mounts of off-the-grid data (bad data). I guess my understanding before was that if the neighborhood size was such that it could not fit entirely inside the domain, then it could not be used for computations. I see that this would not be practical for except for very small neighborhoods. 

Good idea on using the masking region to avoid the inclusion of bad data!

**Reply by JohnHalleyGotway** [reply]
@jwraby yes, you got it. When applying HiRA in Point-Stat, if the neighborhood around each observation is large enough to include any bad data or go off the grid, then that point is excluded from that verification task. Whereas for a smaller neighborhood that doesn't go off the grid, that point would be included.

I could imagine an enhancement to Point-Stat where, rather than processing the obs independently for each neighborhood size (as is done now), we only use the intersection of point observations that appear in all neighborhood sizes. That would make it easier to compare results across multiple neighborhood sizes rather than having to define a "small enough" interior masking region. So basically, the TOTAL column would remain constant across all neighborhood sizes.

Is that the behavior you were expecting in the first place?

**Reply by jwraby** [reply]
I think your approach to restrict the use of observations to ensure that the intersection subset is used, is, by far, much more well thought out and practical than my previous concept ever was! I agree that it seems like a very nice solution to provide a way to compare results across multiple neighborhood sizes. The tradeoff would appear to be that the intersection subset of obs would be a sizeable reduction from the original TOTAL when you include larger neighborhood sizes. For my experiment, the widths ranged from 3 to 49 and the decrease from the TOTAL value for width = 3 to the TOTAL value for width = 49 was about 50%.

**Reply by jwraby** [reply]
Just one additional thought on this. Regarding a possible (and very useful) enhancement along the lines of what you mentioned, I think it would be good to have the option to use the intersection of point observations that appear in all neighborhood sizes in addition to the present behavior. In my experiment I could determine the neighborhood size which resulted in the decrease of the TOTAL value. This was valuable for my purpose which is to use the neighborhood size which does not result in a decrease in the TOTAL value from that which appears in the MPR output.

**Reply by JohnHalleyGotway** [reply]
@mpm-meto and @TaraJensen, as the original source of the HiRA methodology and METplus manager, respectively, can you please take a close look at this discussion topic?

The main issue here is that the larger the neighborhood size you employ, the further the impacts of edge effects encroach into your domain. This is true for HiRA in Point-Stat and Ensemble-Stat and neighborhood methods in Grid-Stat. This is much less of an issue for global domains but greatly affects regional domains.

Users often configure MET to apply multiple neighborhood sizes and see how the performance changes relative to the size. So they typically compare performances ACROSS neighborhood sizes. However, as the size increases, the neighborhoods around more points extend into regions of bad data or off the grid. Typically, the larger the neighborhood size, the smaller the number of matched pairs in the TOTAL column. While comparing across sizes, the number and location of the matched pairs can change.

We've generally just ignored this difference. With the current version of MET, you'd need to choose "small enough" interior masking regions, based on your maximum neighborhood size, to yield a constant value in the TOTAL column. But that's a pretty fine level of detail to ask users to appreciate, may take a lot of trial and error, and depends entirely on the neighborhood widths chosen.

The questions here are whether or not we should enhance MET to make this "event equalization" across neighborhood sizes easy.

I could imagine some new config file flag to control whether or not only the intersection of valid data across all neighborhood sizes should be used, or just process them independently (current behavior). Perhaps...
```
nbrhd_event_equal = TRUE/FALSE; // Set default = FALSE to maintain existing functionality
... and/or...
hira_event_equal = TRUE/FALSE;
```

Practically speaking, I *think* we'd just need to sort the neighborhood sizes and process them largest to smallest rather than in the order they were supplied in the config file. The largest size would produce the minimum amount of valid data and could define our intersection. But we'd need to think through that carefully to make sure the vld_thresh setting doesn't break that assumption.

So the questions to you two are:
1. Is a consistent number of matched pairs in the TOTAL column a worthwhile and important goal for HiRA and neighborhood methods?
2. If so, do these considerations apply equally to HiRA in Point-Stat/Ensemble-Stat as well as neighborhood methods in Grid-Stat? Or are the considerations different in different tools?
3. Should I write up one or more GitHub issues about this?
4. What's the priority?
5. How would we pay for it?

Thanks!

**Reply by JohnHalleyGotway** [reply]
@jwraby, looks like this discussion has stalled out. Do you want me to just mark it as being "answered" and move on. Or do you have any followup discussion on it?
